Source/WebKit/WebProcess/Extensions/Cocoa/_WKWebExtensionWebRequestFilter.mm

```diff
@@ -0,0 +1,248 @@
+        return _WKWebExtensionWebRequestResourceTypeImage;
+    case _WKResourceLoadInfoResourceTypeFont:
+        return _WKWebExtensionWebRequestResourceTypeFont;
+    case _WKResourceLoadInfoResourceTypeObject:
+        return _WKWebExtensionWebRequestResourceTypeObject;
+    case _WKResourceLoadInfoResourceTypeFetch:
+    case _WKResourceLoadInfoResourceTypeXMLHTTPRequest:
+        return _WKWebExtensionWebRequestResourceTypeXMLHTTPRequest;
+    case _WKResourceLoadInfoResourceTypeCSPReport:
+        return _WKWebExtensionWebRequestResourceTypeCSPReport;
+    case _WKResourceLoadInfoResourceTypeMedia:
+        return _WKWebExtensionWebRequestResourceTypeMedia;
+    case _WKResourceLoadInfoResourceTypeApplicationManifest:
+        return _WKWebExtensionWebRequestResourceTypeApplicationManifest;
+    case _WKResourceLoadInfoResourceTypeXSLT:
+        return _WKWebExtensionWebRequestResourceTypeXSLT;
+    case _WKResourceLoadInfoResourceTypePing:
+        return _WKWebExtensionWebRequestResourceTypePing;
+    case _WKResourceLoadInfoResourceTypeBeacon:
+        return _WKWebExtensionWebRequestResourceTypeBeacon;
+    case _WKResourceLoadInfoResourceTypeOther:
+        return _WKWebExtensionWebRequestResourceTypeOther;
+    }
+
+    ASSERT_NOT_REACHED();
+    return _WKWebExtensionWebRequestResourceTypeOther;
+}
+
+@implementation _WKWebExtensionWebRequestFilter {
+    std::optional<WebExtensionTabIdentifier> _tabID;
+    std::optional<WebExtensionWindowIdentifier> _windowID;
+    NSArray<_WKWebExtensionMatchPattern *> *_urlPatterns;
+    NSSet<NSNumber *> *_types;
+}
+
+- (instancetype)initWithDictionary:(NSDictionary<NSString *, id> *)dictionary outErrorMessage:(NSString **)outErrorMessage
+{
+    if (!(self = [super init])) {
+        *outErrorMessage = @"Runtime failure.";
+        return nil;
+    }
+
+    static NSArray<NSString *> *requiredKeys = @[
+        urlsKey,
+    ];
+
+    static NSDictionary<NSString *, id> *expectedTypes = @{
+        urlsKey: @[ NSString.class ],
+        typesKey: NSArray.class,
+        tabIdKey: NSNumber.class,
+        windowIdKey: NSNumber.class,
+    };
+
+    if (!validateDictionary(dictionary, nil, requiredKeys, expectedTypes, outErrorMessage))
+        return nil;
+
+    if ((*outErrorMessage = [self _initializeWithDictionary:dictionary]))
+        return nil;
+
+    return self;
+}
+
+static NSArray<_WKWebExtensionMatchPattern *> *toURLs(NSArray<NSString *> *value, NSString **outErrorMessage)
+{
+    if (!value.count)
+        return nil;
+
+    NSError *error;
+
+    NSMutableArray<_WKWebExtensionMatchPattern *> *patterns = [[NSMutableArray alloc] init];
+    for (NSString *rawPattern in value) {
+        _WKWebExtensionMatchPattern *pattern = [[_WKWebExtensionMatchPattern alloc] initWithString:rawPattern error:&error];
+        if (!pattern) {
+            if (outErrorMessage)
+                *outErrorMessage = error.userInfo[NSDebugDescriptionErrorKey];
```

Comment:
`toErrorString(nil, urlsKey, @"'%@' is an invalid match pattern. %@", rawPattern, error.userInfo[NSDebugDescriptionErrorKey])`